### PR TITLE
[Merged by Bors] - perf(Group.Units): factor out data fields 

### PR DESCRIPTION
**Pull request by mattrobball**
We factor out data fields to avoid unecessary unfolding.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
!bench

File: Mathlib/Algebra/Group/Units.lean [hunk 1]
```diff
@@ -1172,10 +1183,15 @@ section NoncomputableDefs
 
 variable {M : Type*}
 
+-- Unused linter bug at `hM`
```
Comment (urkud) [inline Mathlib/Algebra/Group/Units.lean:1186]: Why do you need to name it, not just write `[Monoid M]`?

File: Mathlib/Algebra/Group/Units.lean [hunk 2]
```diff
@@ -187,14 +187,22 @@ theorem copy_eq (u : αˣ) (val hv inv hi) : u.copy val hv inv hi = u :=
 #align units.copy_eq Units.copy_eq
 #align add_units.copy_eq AddUnits.copy_eq
 
-/-- Units of a monoid form have a multiplication and multiplicative identity. -/
-@[to_additive "Additive units of an additive monoid have an addition and an additive identity."]
-instance instMulOneClass : MulOneClass αˣ where
+/-- Units of a monoid form have an induced multiplication. -/
```
Comment (ScottCarnahan) [inline Mathlib/Algebra/Group/Units.lean:190]: Should `form` be removed?

File: Mathlib/Algebra/Group/Units.lean [hunk 3]
```diff
@@ -187,14 +187,22 @@ theorem copy_eq (u : αˣ) (val hv inv hi) : u.copy val hv inv hi = u :=
...
     ⟨u₁.val * u₂.val, u₂.inv * u₁.inv,
       by rw [mul_assoc, ← mul_assoc u₂.val, val_inv, one_mul, val_inv],
       by rw [mul_assoc, ← mul_assoc u₁.inv, inv_val, one_mul, inv_val]⟩
+
+/-- Units of a monoid have a unit -/
+@[to_additive "Additive units of an additive monoid have a zero."]
+instance : One αˣ where
   one := ⟨1, 1, one_mul 1, one_mul 1⟩
+
+/-- Units of a monoid form have a multiplication and multiplicative identity. -/
```
Comment (ScottCarnahan) [inline Mathlib/Algebra/Group/Units.lean:203]: I think `form` should be removed here too.

File: Mathlib/Algebra/Group/Units.lean (hunk 1 above)
Comment (mattrobball) [reply Mathlib/Algebra/Group/Units.lean:1186]: Done

File: Mathlib/Algebra/Group/Units.lean (hunk 2 above)
Comment (mattrobball) [reply Mathlib/Algebra/Group/Units.lean:190]: Yes! Thanks

File: Mathlib/Algebra/Group/Units.lean (hunk 3 above)
Comment (mattrobball) [reply Mathlib/Algebra/Group/Units.lean:203]: Yes. Thanks!

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**